### PR TITLE
Infrastructure upgrade

### PR DESCRIPTION
**Pull request by mtscout6**
A lot has happened since this projects inception and newer tools are available that can facilitate a better development environment. I have started some work to give the project structure a facelift to meet the current trends of the broader React community. My goals:

- Source cleanup
  - [x] Rename all js files to *.js
  - [x] Use [Babel](http://babeljs.io/) for our es6 -> es5 transpilation
  - Update all js to es6 standards
    - [x] src
    - [x] test
    - [x] docs
    - [x] doc examples
    - [x] ie8 visual test
  - ~~Use [Flow](http://flowtype.org/) for static type checking??~~ (Striking for now, can be another PR)
  - [x] [ESLint](http://eslint.org/) all .js files
- Build
  - [x] Use [Webpack](http://webpack.github.io/) - This should provide more control on how we build our pre-concatenated and minified releases.
  - [x] Drop Grunt, use npm scripts and shell scripts where needed.
  - [x] Promise based release process (faster)
  - [x] Update README to explain how to run.
- Test
  - [x] Single karma config file
  - [x] Not dependent on building src to disk to run, use [karma-webpack](https://github.com/webpack/karma-webpack) which will build the src in memory.
  - [x] Better test watcher that re-runs tests when changing src files. Current watcher only runs if test files change.
  - [x] Fail on linting errors
  - [x] Pull `es5-shim` from npm
- Docs
  - [x] Pull `bootstrap` from npm
  - [x] Pull `codemirror` from npm
  - [x] Use in browser Babel transformer for doc examples
  - [x] Consolidate package.json deps and scripts to repo root. Leads to less confusion on how to run the docs locally.
  - ~~Change router to react-router, which tends to be up to date with React quicker.~~ (Taken care of in #424)
  - [x] Update README for local setup.

This is still a work in progress so I'm open to suggestions for improvement. __I am planning to continually rebase and force push this branch, until it's complete.__

### CODE REVIEW

**Comment by mtscout6** [comment]
I'm ready to declare success on this. Does anyone see anything glaring that should be addressed before I bring this into master? I'd like to bring this in by end of day tomorrow, baring any revelations by the community.

/cc @joemcbride @pieterv @stevoland @trentgrover-wf

**Comment by joemcbride** [comment]
Really glad to see this update.  :+1: 

I pulled down the source and have been doing some smoke testing.  I'm seeing some funky styling /scrolling issues with CodeMirror.  I also ran into an issue with eslint not finding a rule (I have latest eslint).  Double checking that its not a local issue.

**Comment by joemcbride** [comment]
So I cleared out all my local npm modules and re-ran npm install.  All unit tests pass.

Eslint dies on: `Error: Definition for rule 'no-global-strict' was not found.`

This is what I'm seeing in the docs in Chrome/Safari/Firefox:
screen shot 2015-03-25 at 11 21 33 am

The re-compile of code works fine, just funky styling.

File: docs/src/CodeMirror.css [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+.highlight, .code-toggle.open {
+  background-color: #fdf6e3;
+}
+
+.cm-s-solarized .cm-comment { color: #93a1a1; }
+
+.CodeMirror {
+  height: auto;
```
Comment (joemcbride) [inline docs/src/CodeMirror.css]: Looks like this line is the cause of the extra-tall editor height, through removing it causes other issues (scrolbars).

Comment (mtscout6) [inline docs/src/CodeMirror.css]: Ya, we fixed this on the last pull request that came in and I forgot to apply it here. I need to do this: https://github.com/react-bootstrap/react-bootstrap/blob/master/docs/assets/style.css#L71-L73. Thanks for the find, I'll update that.

**Comment by mtscout6** [comment]
I ran into that eslint issue as well at one point. I made sure that everything was up to date and it went away. Is it likely that one of your global eslint installs is causing problems? Do you see it when you run `npm run lint`?

File: docs/src/CodeMirror.css (hunk 1 above)
Comment (joemcbride) [inline docs/src/CodeMirror.css]: :+1:

**Comment by joemcbride** [comment]
I may not have the global react plugin mac side.  Checking that.

**Comment by mtscout6** [comment]
It should just work if you do it through `npm run lint` is that the case for you?

**Comment by joemcbride** [comment]
No, I get the error using `npm run lint`.

**Comment by joemcbride** [comment]
```
$ npm run lint

> react-bootstrap@0.18.0 lint /Users/jomc/Dev/rb-es6
> eslint src test docs ie8 tools webpack karma.conf.js webpack.config.js webpack.docs.js

/Users/jomc/Dev/rb-es6/node_modules/eslint/lib/eslint.js:657
                    throw new Error("Definition for rule '" + key + "' was not
                          ^
Error: Definition for rule 'no-global-strict' was not found.
    at /Users/jomc/Dev/rb-es6/node_modules/eslint/lib/eslint.js:657:27
    at Array.forEach (native)
    at EventEmitter.module.exports.api.verify (/Users/jomc/Dev/rb-es6/node_modules/eslint/lib/eslint.js:630:16)
    at processFile (/Users/jomc/Dev/rb-es6/node_modules/eslint/lib/cli-engine.js:193:27)
    at /Users/jomc/Dev/rb-es6/node_modules/eslint/lib/cli-engine.js:293:26
    at /Users/jomc/Dev/rb-es6/node_modules/eslint/lib/util/traverse.js:61:17
    at Array.forEach (native)
    at traverse (/Users/jomc/Dev/rb-es6/node_modules/eslint/lib/util/traverse.js:41:54)
    at /Users/jomc/Dev/rb-es6/node_modules/eslint/lib/util/traverse.js:63:17
    at Array.forEach (native)
```

**Comment by joemcbride** [comment]
Since CI passes, I can only guess this is a local issue for me.  I would be fine with you pulling this in while I figure this out.

**Comment by mtscout6** [comment]
Turns out the eslint problem for @joemcbride was with the `docs/node_modules` folder. For those of you that have had this repo cloned before it would be a good idea to run `git clean -dfx` after this PR is merged and re-run npm install. If that does not completely clear your `node_modules` and `docs/node_modules` folders then you will need to remove them manually.

**Pull request merged**